### PR TITLE
Update ensemble to handle decoupled request/response

### PR DESCRIPTION
**Pull request by GuanLuo**

### CODE REVIEW

File: src/core/ensemble_scheduler.cc [hunk 1]
```diff
@@ -41,22 +41,57 @@ namespace {
 
 class EnsembleContext;
 
+using IterationCount = size_t;
```
Comment (CoderHam) [inline src/core/ensemble_scheduler.cc:44]: Why is this necessary?

Comment (GuanLuo) [reply src/core/ensemble_scheduler.cc:44]: For clarity, too many `size_t` for different kinds of variables.

**Comment by GuanLuo** [comment]
@CoderHam @deadeyegoodwin @tanmayv25 The PR is ready for review
